tests/assembly/simd-intrinsic-select.rs

```diff
@@ -0,0 +1,99 @@
+//@ [x86-avx2] needs-llvm-components: x86
+//@ [x86-avx512] compile-flags: --target=x86_64-unknown-linux-gnu -C llvm-args=-x86-asm-syntax=intel
+//@ [x86-avx512] compile-flags: -C target-feature=+avx512f,+avx512vl,+avx512bw,+avx512dq
+//@ [x86-avx512] needs-llvm-components: x86
+//@ [aarch64] compile-flags: --target=aarch64-unknown-linux-gnu
+//@ [aarch64] needs-llvm-components: aarch64
+//@ [aarch64] min-llvm-version: 18.0
+//@ assembly-output: emit-asm
+//@ compile-flags: --crate-type=lib -O
+
+#![feature(no_core, lang_items, repr_simd, intrinsics)]
+#![no_core]
+#![allow(non_camel_case_types)]
+
+// Because we don't have core yet.
+#[lang = "sized"]
+pub trait Sized {}
+
+#[lang = "copy"]
+trait Copy {}
+
+#[repr(simd)]
+pub struct i8x16([i8; 16]);
+
+#[repr(simd)]
+pub struct m8x16([i8; 16]);
+
+#[repr(simd)]
+pub struct f32x8([f32; 8]);
+
+#[repr(simd)]
+pub struct m32x8([i32; 8]);
+
+#[repr(simd)]
+pub struct f64x4([f64; 4]);
+
+#[repr(simd)]
+pub struct m64x4([i64; 4]);
+
+extern "rust-intrinsic" {
+    fn simd_select<M, V>(mask: M, a: V, b: V) -> V;
+}
+
+// CHECK-LABEL: select_i8x16
+#[no_mangle]
+pub unsafe extern "C" fn select_i8x16(mask: m8x16, a: i8x16, b: i8x16) -> i8x16 {
+    // x86-avx2: vpsllw xmm0, xmm0, 7
+    // x86-avx2-NEXT: vpblendvb xmm0, xmm2, xmm1, xmm0
+    //
+    // x86-avx512: vpsllw xmm0, xmm0, 7
+    // x86-avx512-NEXT: vpmovb2m k1, xmm0
+    // x86-avx512-NEXT: vpblendmb xmm0 {k1}, xmm2, xmm1
+    //
+    // aarch64: shl v0.16b, v0.16b, #7
+    // aarch64-NEXT: cmlt v0.16b, v0.16b, #0
+    // aarch64-NEXT: bsl v0.16b, v1.16b, v2.16b
+    simd_select(mask, a, b)
+}
+
+// CHECK-LABEL: select_f32x8
+#[no_mangle]
+pub unsafe extern "C" fn select_f32x8(mask: m32x8, a: f32x8, b: f32x8) -> f32x8 {
+    // x86-avx2: vpslld ymm0, ymm0, 31
+    // x86-avx2-NEXT: vblendvps ymm0, ymm2, ymm1, ymm0
+    //
+    // x86-avx512: vpslld ymm0, ymm0, 31
+    // x86-avx512-NEXT: vpmovd2m k1, ymm0
+    // x86-avx512-NEXT: vblendmps ymm0 {k1}, ymm2, ymm1
+    //
+    // aarch64: shl v0.4s, v0.4s, #31
+    // aarch64-NEXT: shl v1.4s, v1.4s, #31
+    // aarch64-NEXT: cmlt v0.4s, v0.4s, #0
+    // aarch64-NEXT: cmlt v1.4s, v1.4s, #0
+    // aarch64-NEXT: bsl v0.16b, v2.16b, v4.16b
+    // aarch64-NEXT: bsl v1.16b, v3.16b, v5.16b
+    simd_select(mask, a, b)
+}
+
+// CHECK-LABEL: select_f64x4
+#[no_mangle]
+pub unsafe extern "C" fn select_f64x4(mask: m64x4, a: f64x4, b: f64x4) -> f64x4 {
+    // x86-avx2: vpsllq ymm0, ymm0, 63
+    // x86-avx2-NEXT: vblendvpd ymm0, ymm2, ymm1, ymm0
+    //
+    // x86-avx512: vpsllq ymm0, ymm0, 63
+    // x86-avx512-NEXT: vpmovq2m k1, ymm0
+    // x86-avx512-NEXT: vblendmpd ymm0 {k1}, ymm2, ymm1
+    //
+    // aarch64: shl v0.2d, v0.2d, #63
+    // aarch64-NEXT: shl v1.2d, v1.2d, #63
+    // aarch64-NEXT: cmlt v0.2d, v0.2d, #0
+    // aarch64-NEXT: cmlt v1.2d, v1.2d, #0
+    // aarch64-NEXT: bsl v0.16b, v2.16b, v4.16b
+    // aarch64-NEXT: bsl v1.16b, v3.16b, v5.16b
+    simd_select(mask, a, b)
+}
```

Comment:
I'm sorry, I wasn't thinking clearly when I read these tests the first time, it seems. These shouldn't test aarch64 at all, unfortunately. The tests on aarch64 should only be for 128-bit vector operations. The output that LLVM is giving may be a sensible interpretation of the request we are making to it, but it is not "in-spec" for `extern "C"`, so this test's output is meaningless for that architecture.

Reply:
That, or we should find a way to avoid using `extern "C"` here, or comment **heavily** on what is going on and our actual expectations.

Reply:
One option is using [FileCheck's string-capture](https://llvm.org/docs/CommandGuide/FileCheck.html#filecheck-string-substitution-blocks) so that we can verify a register is reused, instead of caring about which register it is.

Reply:
Ah, the C abi does not define how to pass 256-bit vectors on arm? Bummer, but kind of obvious when you spell it out :) And would also affect sse tests.

If I remember correctly how the checks looked with Rust abi, then vectors are passed by pointer, but are the registers used for these pointers defined as strictly as in the C abi (first arg in rdi, 2nd in rsi, rcx, rdx)?

I think I would rather add a separate function with f64x2 / f32x4 args for checking aarch64 and sse.

Reply:
What it _does_ define is how to pass this type:
https://doc.rust-lang.org/stable/core/arch/arm/struct.float32x4x2_t.html

Reply:
No, the Rust ABI does not strictly define those registers.